skgstat/DirectionalVariogram.py

```diff
@@ -277,6 +281,39 @@ def __init__(self,
         # Note that fit() calls preprocessing
         self.fit(force=True)
 
+    def preprocessing(self, force=False):
+        self._calc_distances(force=force)
+        self._calc_direction_mask_data(force)
+        self._calc_diff(force=force)
+        self._calc_groups(force=force)
+
+    def _calc_direction_mask_data(self, force=False):
+        if self._angles is not None and not force:
+            return
+
+        # if self._X is of just one dimension, concat zeros.
+        if self._X.ndim == 1:
+            _x = np.vstack(zip(self._X, np.zeros(len(self._X))))
+        else:
+            _x = self._X
+
+        if self._dist_func == "euclidean":
+            self._euclidean_dist = self._dist
+        else:
+            self._euclidean_dist = scipy.spatial.distance.pdist(_x, "euclidean")
+
+        # Calculate the angle between the vector between the two
+        # points, and east (see comment about self.azimuth)
+        # Note that this is in radians, while azimuth is in degrees.
+        # (a - b).[1,0] = ||a - b|| * ||[1,0]|| * cos(v)
+        # cos(v) = (a - b).[1,0] / ||a - b||
+        # cos(v) = (a.[1,0] - b.[1,0]) / ||a - b||
+        pos_angles = np.arccos(scipy.spatial.distance.pdist(np.array([np.dot(_x, [1, 0])]).T, np.subtract) / self._euclidean_dist)
+        # cos(v) for [2,1] and [2, -1] is the same, but v is not (v vs -v), fix that.
+        ydiff = scipy.spatial.distance.pdist(np.array([np.dot(_x, [0, 1])]).T, np.subtract)
+        self._angles = np.where(ydiff >= 0, pos_angles, -pos_angles)
```

Comment:
Hey @redhog ,
I think I have some trouble to understand this part, maybe you can help me to sort things out. This function is calculating the angle between the vector of a point pair and the x-axis, right?

I ran a short test to calculate some angles. Take the points `A = (0,0)`, `B = (1,2)` and `C = (1,2)`. If I pass them to the variogram, the condensed angles matrix `DirectionalVariogram._angles` should contain the angles of `AB` and `AC` to the x-axis as the first two elements.
I ran this code to do that:
```python
import numpy as np
from skgstat import DirectionalVariogram 

a = np.array([[0,0], [1,2], [2,1]])

# trying to run fit( ) causes an error for only 3 points, so I overwrite the method
class V(DirectionalVariogram):
  def fit(*args, **kwargs):
    pass

DV = V(a, np.random.normal(0,1,size=3))
DV._calc_direction_mask_data()
np.degrees(DV._angles)[:2]
```
and I get this result:
```
array([-116.56505118, -153.43494882])
```

But the results should be `~63°` for the angle between the vectors `[1,0]` and `[1,2]` , which would be the first element and `~26°` for the second.

I am not sure, if I don't get the function right or if I don't get the math right or if there is an error in our function.
With pen and paper I calculated for the first point:
Let `a` be the angle needed and `v = AB = [1,2]` and `e = [1, 0]` :
`cos(a) = (v . e) / ( |v| * |e| )`  (where `v . e` would be the scalar product)
then:
`cos(a) = (1*1 + 2* 0) / (sqrt(5) * sqrt(1)) = 1 / sqrt(5)`
and 
`a = arccos(1 / sqrt(5)) = 1.107 rad    =   1.107 * 180 / pi    =   63.4...°`

Am I blind? Do you see the mistake?

Reply:
```
>>> np.degrees(DV._angles)[:2] + 180
array([63.43494882, 26.56505118])
```

`scipy.spatial.distance.pdist` assumes distances are commutative and takes some shortcuts based on that that basically swaps the two points of a point pair sometimes. That means that the angles in `angles` are only correct +- 180 degrees, but since our filtering is symmetrical, I thought using pdist this way would be ok.

So no, you're not blind :)